File: src/simulation/structure.ts
import type { AuthoredSurface, ShipDefinition, Vec3 } from '../ships/blueprint';
import { contains, segmentBox, sub } from './geometry';
import { segmentPlate } from './protection';

// Reconstructed exterior belt envelopes can sit about 1.1 m inside the faired
// hull. Treat that skin and its belt as one exterior layer in combat and review.
export const EXTERIOR_PLATING_REPLACEMENT_M = 1.5;

export interface StructuralSurface extends AuthoredSurface {
  id: string; name: string; thicknessMm: number; hull: boolean; center: Vec3; size: Vec3;
  chunks: { center: Vec3; size: Vec3; first: number; end: number }[];
}
const cache=new WeakMap<ShipDefinition,StructuralSurface[]>();
const bounds=(vertices:Vec3[])=>{
  const low=[0,1,2].map(i=>Math.min(...vertices.map(v=>v[i]))), high=[0,1,2].map(i=>Math.max(...vertices.map(v=>v[i])));
  return {center:low.map((v,i)=>(v+high[i])/2) as Vec3,size:low.map((v,i)=>Math.max(.00001,high[i]-v)) as Vec3};
};

/** Ear clipping preserves concave deckhouse footprints, including bridge wings. */
function cap(points:[number,number][]):[number,number,number][] {
  const cross=(a:number,b:number,c:number)=>(points[b][0]-points[a][0])*(points[c][1]-points[a][1])-(points[b][1]-points[a][1])*(points[c][0]-points[a][0]);
  const area=points.reduce((a,p,i)=>a+p[0]*points[(i+1)%points.length][1]-p[1]*points[(i+1)%points.length][0],0);
  const ids=points.map((_,i)=>i);if(area<0)ids.reverse();
  const result:[number,number,number][]=[];
  while(ids.length>2) {
    let found=false;
    for(let i=0;i<ids.length;i++) {
      const a=ids[(i+ids.length-1)%ids.length],b=ids[i],c=ids[(i+1)%ids.length];
      if(Math.abs(cross(a,b,c))<1e-9){ids.splice(i,1);found=true;break;}
      if(cross(a,b,c)<0 || ids.some(p=>p!==a&&p!==b&&p!==c&&cross(a,b,p)>=-1e-9&&cross(b,c,p)>=-1e-9&&cross(c,a,p)>=-1e-9))continue;
      result.push([a,b,c]);ids.splice(i,1);found=true;break;
    }
    if(!found)throw new Error('Structural footprint cannot be triangulated');
  }
  return result;
}

/** Generated once from original blueprint surfaces, never from the render mesh. */
export function structuralSurfaces(def:ShipDefinition):StructuralSurface[] {
  const cached=cache.get(def);if(cached)return cached;
  const plating=def.structuralPlating;if(!plating)return [];
  const result:StructuralSurface[]=[];
  function body(id:string,name:string,surface:AuthoredSurface,thicknessMm:number,hull=false) {
    const triangles=surface.triangles.filter(ids=>{
      const [a,b,c]=ids.map(i=>surface.vertices[i]),u=sub(b,a),v=sub(c,a);
      return Math.hypot(u[1]*v[2]-u[2]*v[1],u[2]*v[0]-u[0]*v[2],u[0]*v[1]-u[1]*v[0])>1e-9;
    });
    const chunks:StructuralSurface['chunks']=[];
    for(let first=0;first<triangles.length;first+=96){const end=Math.min(first+96,triangles.length);chunks.push({...bounds(triangles.slice(first,end).flatMap(t=>t.map(i=>surface.vertices[i]))),first,end});}
    result.push({id,name,...surface,triangles,thicknessMm,hull,...bounds(surface.vertices),chunks});
  }
  const sections=def.hull.sections!, n=sections[0].points.length*2-1;
  const vertices=sections.flatMap(s=>[...s.points,...s.points.slice(1).reverse().map(([w,y])=>[-w,y])].map(([w,y])=>[-w,y,def.hull.length/2-s.station] as Vec3));
  const triangles:[number,number,number][]=[];
  for(let s=0;s<sections.length-1;s++)for(let j=0;j<n;j++){
    const a=s*n+j,b=s*n+(j+1)%n,c=(s+1)*n+(j+1)%n,d=(s+1)*n+j;
    // A recurved stem can have a zero-breadth interval above its bulb.
    // Opposite coincident faces there are air, not a centerline steel sheet.
    const halves=j>=sections[0].points.length?[[a,b,d],[b,c,d]]:[[a,b,c],[a,c,d]];
    for(const ids of halves as [number,number,number][])
      if(ids.some(i=>Math.abs(vertices[i][0])>1e-7))triangles.push(ids);
  }
  // A broad transom is an exterior face too. Pointed ends naturally reduce to
  // zero-area caps; ear clipping drops their collapsed/collinear vertices.
  for(const section of [0,sections.length-1]) {
    const offset=section*n,points=vertices.slice(offset,offset+n).map(v=>[v[0],v[1]] as [number,number]);
    for(const [a,b,c] of cap(points))triangles.push(section===0?[offset+a,offset+b,offset+c]:[offset+c,offset+b,offset+a]);
  }
  body('hull','Hull shell · bow to stern',{vertices,triangles},plating.hullMm,true);
  for(const s of def.structures??[]) {
    if(s.surface){body(s.id,s.name,s.surface,plating.superstructureMm);continue;}
    const n=s.footprint.length,vertices=[s.baseY,s.baseY+s.height].flatMap(y=>s.footprint.map(([x,z])=>[x,y,z] as Vec3));
    const triangles:[number,number,number][]=[];
    for(let i=0;i<n;i++){const j=(i+1)%n;triangles.push([i,j,n+j],[i,n+j,n+i]);}
    for(const [a,b,c] of cap(s.footprint))triangles.push([a,b,c],[n+a,n+c,n+b]);
    body(s.id,s.name,{vertices,triangles},plating.superstructureMm);
  }
  cache.set(def,result);return result;
}

export function structuralHits(from:Vec3,to:Vec3,def:ShipDefinition) {
  return structuralSurfaces(def).flatMap(surface=>{
    if(!segmentBox(from,to,surface))return [];
    const hits=[];
    for(const chunk of surface.chunks){
      if(!segmentBox(from,to,chunk))continue;
      for(let i=chunk.first;i<chunk.end;i++){
        const hit=segmentPlate(from,to,surface.triangles[i].map(j=>surface.vertices[j]));
        if(hit)hits.push({...hit,surface,triangle:i});
      }
    }
    return hits;
  }).sort((a,b)=>a.t-b.t || a.surface.id.localeCompare(b.surface.id) || a.triangle-b.triangle);
}

/** The sea cannot occupy the intact hull interior, regardless of its armor layout. */
export function insideHull(point:Vec3,def:ShipDefinition):boolean {
  if(!def.structuralPlating)return def.armor.some(a=>!a.plate&&contains(a,point));
  const station=def.hull.length/2-point[2],sections=def.hull.sections!;
  const index=sections.findIndex((s,i)=>i<sections.length-1&&station>=s.station&&station<=sections[i+1].station);
  if(index<0)return false;
  const a=sections[index],b=sections[index+1],t=(station-a.station)/(b.station-a.station);
  const points=a.points.map(([w,y],i)=>[w+(b.points[i][0]-w)*t,y+(b.points[i][1]-y)*t]);
  if(point[1]<points[0][1]-1e-6||point[1]>points.at(-1)![1]+1e-6)return false;
  let width=-1;
  for(let i=0;i<points.length-1;i++){
    const [wa,ya]=points[i],[wb,yb]=points[i+1];
    if(point[1]>=ya-1e-6&&point[1]<=yb+1e-6)width=Math.max(width,Math.abs(yb-ya)<1e-8?Math.max(wa,wb):wa+(wb-wa)*(point[1]-ya)/(yb-ya));
  }
  return width>1e-7&&Math.abs(point[0])<=width+1e-6;
}

File: crates/naval-sim/src/structure.rs
use crate::{
    definition::{ShipDefinition, Vec3},
    geometry::*,
    protection::{PlateHit, segment_plate},
};
pub const EXTERIOR_PLATING_REPLACEMENT_M: f64 = 1.5;
#[derive(Clone, Debug)]
pub struct StructuralSurface {
    pub id: String,
    pub name: String,
    pub thickness_mm: f64,
    pub hull: bool,
    pub vertices: Vec<Vec3>,
    pub triangles: Vec<[usize; 3]>,
    center: Vec3,
    size: Vec3,
    chunks: Vec<Chunk>,
}
#[derive(Clone, Debug)]
struct Chunk {
    center: Vec3,
    size: Vec3,
    first: usize,
    end: usize,
}
pub fn bounds(points: impl Iterator<Item = Vec3>) -> (Vec3, Vec3) {
    let (mut low, mut high) = ([f64::INFINITY; 3], [f64::NEG_INFINITY; 3]);
    for p in points {
        for i in 0..3 {
            low[i] = low[i].min(p[i]);
            high[i] = high[i].max(p[i]);
        }
    }
    (
        std::array::from_fn(|i| (low[i] + high[i]) / 2.0),
        std::array::from_fn(|i| (high[i] - low[i]).max(0.00001)),
    )
}
fn cap(points: &[[f64; 2]]) -> Result<Vec<[usize; 3]>, String> {
    let cross = |a: usize, b: usize, c: usize| {
        (points[b][0] - points[a][0]) * (points[c][1] - points[a][1])
            - (points[b][1] - points[a][1]) * (points[c][0] - points[a][0])
    };
    let area: f64 = points
        .iter()
        .enumerate()
        .map(|(i, p)| {
            p[0] * points[(i + 1) % points.len()][1] - p[1] * points[(i + 1) % points.len()][0]
        })
        .sum();
    let mut ids: Vec<_> = (0..points.len()).collect();
    if area < 0.0 {
        ids.reverse();
    }
    let mut result = vec![];
    while ids.len() > 2 {
        let mut found = false;
        for i in 0..ids.len() {
            let (a, b, c) = (
                ids[(i + ids.len() - 1) % ids.len()],
                ids[i],
                ids[(i + 1) % ids.len()],
            );
            if cross(a, b, c).abs() < 1e-9 {
                ids.remove(i);
                found = true;
                break;
            }
            if cross(a, b, c) < 0.0
                || ids.iter().any(|&p| {
                    p != a
                        && p != b
                        && p != c
                        && cross(a, b, p) >= -1e-9
                        && cross(b, c, p) >= -1e-9
                        && cross(c, a, p) >= -1e-9
                })
            {
                continue;
            }
            result.push([a, b, c]);
            ids.remove(i);
            found = true;
            break;
        }
        if !found {
            return Err("structural footprint cannot be triangulated".into());
        }
    }
    Ok(result)
}
impl StructuralSurface {
    fn new(
        id: String,
        name: String,
        vertices: Vec<Vec3>,
        triangles: Vec<[usize; 3]>,
        thickness_mm: f64,
        hull: bool,
    ) -> Self {
        let triangles: Vec<_> = triangles
            .into_iter()
            .filter(|ids| {
                length(cross(
                    sub(vertices[ids[1]], vertices[ids[0]]),
                    sub(vertices[ids[2]], vertices[ids[0]]),
                )) > 1e-9
            })
            .collect();
        let chunks = (0..triangles.len())
            .step_by(96)
            .map(|first| {
                let end = (first + 96).min(triangles.len());
                let (center, size) = bounds(
                    triangles[first..end]
                        .iter()
                        .flat_map(|t| t.map(|i| vertices[i])),
                );
                Chunk {
                    center,
                    size,
                    first,
                    end,
                }
            })
            .collect();
        let (center, size) = bounds(vertices.iter().copied());
        Self {
            id,
            name,
            vertices,
            triangles,
            thickness_mm,
            hull,
            chunks,
            center,
            size,
        }
    }
}
pub fn structural_surfaces(def: &ShipDefinition) -> Result<Vec<StructuralSurface>, String> {
    let Some(plating) = &def.structural_plating else {
        return Ok(vec![]);
    };
    let sections = def
        .hull
        .sections
        .as_ref()
        .ok_or("structural plating requires sections")?;
    let first = sections.first().ok_or("empty hull sections")?;
    let n = first.points.len() * 2 - 1;
    if sections
        .iter()
        .any(|s| s.points.len() != first.points.len())
    {
        return Err("structural sections require matching point counts".into());
    }
    let vertices: Vec<Vec3> = sections
        .iter()
        .flat_map(|s| {
            s.points
                .iter()
                .copied()
                .chain(s.points.iter().skip(1).rev().map(|p| [-p[0], p[1]]))
                .map(|p| [-p[0], p[1], def.hull.length / 2.0 - s.station])
        })
        .collect();
    let mut triangles = vec![];
    for s in 0..sections.len() - 1 {
        for j in 0..n {
            let (a, b, c, d) = (
                s * n + j,
                s * n + (j + 1) % n,
                (s + 1) * n + (j + 1) % n,
                (s + 1) * n + j,
            );
            let halves = if j >= first.points.len() {
                [[a, b, d], [b, c, d]]
            } else {
                [[a, b, c], [a, c, d]]
            };
            for ids in halves {
                if ids.iter().any(|&i| vertices[i][0].abs() > 1e-7) {
                    triangles.push(ids);
                }
            }
        }
    }
    for section in [0, sections.len() - 1] {
        let offset = section * n;
        let points: Vec<_> = vertices[offset..offset + n]
            .iter()
            .map(|v| [v[0], v[1]])
            .collect();
        for [a, b, c] in cap(&points)? {
            triangles.push(if section == 0 {
                [offset + a, offset + b, offset + c]
            } else {
                [offset + c, offset + b, offset + a]
            });
        }
    }
    let mut result = vec![StructuralSurface::new(
        "hull".into(),
        "Hull shell · bow to stern".into(),
        vertices,
        triangles,
        plating.hull_mm,
        true,
    )];
    for s in def.structures.iter().flatten() {
        if let Some(surface) = &s.surface {
            result.push(StructuralSurface::new(
                s.id.clone(),
                s.name.clone(),
                surface.vertices.clone(),
                surface
                    .triangles
                    .iter()
                    .map(|t| t.map(|i| i as usize))
                    .collect(),
                plating.superstructure_mm,
                false,
            ));
            continue;
        }
        let n = s.footprint.len();
        let vertices = [s.base_y, s.base_y + s.height]
            .into_iter()
            .flat_map(|y| s.footprint.iter().map(move |p| [p[0], y, p[1]]))
            .collect();
        let mut triangles = vec![];
        for i in 0..n {
            let j = (i + 1) % n;
            triangles.extend([[i, j, n + j], [i, n + j, n + i]]);
        }
        for [a, b, c] in cap(&s.footprint)? {
            triangles.extend([[a, b, c], [n + a, n + c, n + b]]);
        }
        result.push(StructuralSurface::new(
            s.id.clone(),
            s.name.clone(),
            vertices,
            triangles,
            plating.superstructure_mm,
            false,
        ));
    }
    Ok(result)
}
#[derive(Clone, Debug)]
pub struct StructuralHit<'a> {
    pub hit: PlateHit,
    pub surface: &'a StructuralSurface,
    pub triangle: usize,
}
pub fn structural_hits(
    from: Vec3,
    to: Vec3,
    surfaces: &[StructuralSurface],
) -> Vec<StructuralHit<'_>> {
    let mut hits = vec![];
    for surface in surfaces {
        if segment_box(from, to, surface.center, surface.size).is_none() {
            continue;
        }
        for chunk in &surface.chunks {
            if segment_box(from, to, chunk.center, chunk.size).is_none() {
                continue;
            }
            for i in chunk.first..chunk.end {
                if let Some(hit) =
                    segment_plate(from, to, &surface.triangles[i].map(|j| surface.vertices[j]))
                {
                    hits.push(StructuralHit {
                        hit,
                        surface,
                        triangle: i,
                    });
                }
            }
        }
    }
    hits.sort_by(|a, b| {
        a.hit
            .t
            .total_cmp(&b.hit.t)
            .then_with(|| a.surface.id.cmp(&b.surface.id))
            .then_with(|| a.triangle.cmp(&b.triangle))
    });
    hits
}
